File: packages/build-spec/test/resolution_test.exs
defmodule FrameshiftBuild.ResolutionTest do
  @moduledoc false

  use ExUnit.Case, async: true

  @profile File.read!(Path.join(__DIR__, "fixtures/profile-v1.json"))
  @plan File.read!(Path.join(__DIR__, "fixtures/assembly-v1.json"))

  test "resolves all exact content pins independently of input order" do
    {plan, profiles, _} = fixture()
    assert {:ok, result} = FrameshiftBuild.resolve_build(plan, profiles)
    assert {:ok, ^result} = FrameshiftBuild.resolve_build(plan, Enum.reverse(profiles))
    assert {:ok, result.identity} == FrameshiftBuild.build_identity(plan)
    assert {:resolution, _, [_, _], []} = result.resolution
  end

  test "missing profiles remain unresolved instead of following a newer revision" do
    {plan, [first, _], [_, missing]} = fixture()
    assert {:ok, result} = FrameshiftBuild.resolve_build(plan, [first])
    assert {:resolution, _, [_], [^missing]} = result.resolution
    assert {:ok, empty} = FrameshiftBuild.resolve_build(plan, [])
    assert {:resolution, _, [], [_, _]} = empty.resolution
    newer = String.replace(first, "\"part_revision\":\"R1\"", "\"part_revision\":\"R2\"")
    assert {:error, "unreferenced_profile"} = FrameshiftBuild.resolve_build(plan, [newer])
  end

  test "tampering, duplicate bodies and caller-supplied digests cannot bind pins" do
    {plan, [first, second], [pin, _]} = fixture()
    assert {:error, "duplicate_identifier"} = FrameshiftBuild.resolve_build(plan, [first, first])
    assert {:error, "noncanonical"} = FrameshiftBuild.resolve_build(plan, [first <> " ", second])

    assert {:error, "invalid_document"} =
             FrameshiftBuild.resolve_build(plan, [%{identity: pin, canonical: second}])

    changed = String.replace(first, "\"max\":1010", "\"max\":1011")

    assert {:error, "unreferenced_profile"} =
             FrameshiftBuild.resolve_build(plan, [changed, second])
  end

  test "type, count, document and total byte limits fail before hashing" do
    for profiles <- [nil, %{}, [nil], [@profile | :improper]] do
      assert {:error, "invalid_document"} = FrameshiftBuild.resolve_build(@plan, profiles)
    end

    assert {:error, "invalid_count"} =
             FrameshiftBuild.resolve_build(@plan, List.duplicate("", 65))

    assert {:error, "too_large"} =
             FrameshiftBuild.resolve_build(@plan, [String.duplicate(" ", 262_145)])

    assert {:error, "too_large"} =
             FrameshiftBuild.resolve_build(
               @plan,
               List.duplicate(String.duplicate(" ", 262_144), 16)
             )

    assert {:error, "invalid_document"} = FrameshiftBuild.resolve_build(nil, [])
  end

  defp fixture do
    second = String.replace(@profile, "\"id\":\"fixture-0\"", "\"id\":\"fixture-1\"")
    {:ok, first_id} = FrameshiftBuild.profile_identity(@profile)
    {:ok, second_id} = FrameshiftBuild.profile_identity(second)

    plan =
      @plan
      |> String.replace("sha256:" <> String.duplicate("a", 64), first_id)
      |> String.replace("sha256:" <> String.duplicate("b", 64), second_id)

    {plan, [@profile, second], [first_id, second_id]}
  end
end

File: packages/build-spec/test/resolution.test.mjs
import {readFile} from 'node:fs/promises';
import {test} from 'node:test';
import assert from 'node:assert/strict';
import {profileIdentity} from '../js/profile.mjs';
import {resolveBuild} from '../js/build.mjs';
const first=await readFile(new URL('./fixtures/profile-v1.json',import.meta.url),'utf8');
const second=first.replace('"id":"fixture-0"','"id":"fixture-1"');
const firstId=(await profileIdentity(first)).identity;
const secondId=(await profileIdentity(second)).identity;
const plan=(await readFile(new URL('./fixtures/assembly-v1.json',import.meta.url),'utf8')).replace('sha256:'+'a'.repeat(64),firstId).replace('sha256:'+'b'.repeat(64),secondId);

test('exact content resolution preserves order independence and missing profiles',async()=>{
  const result=await resolveBuild(plan,[first,second]);
  assert.equal(result.ok,true);
  assert.deepEqual(result,await resolveBuild(plan,[second,first]));
  assert.deepEqual(result.resolution.missing.toArray(),[]);
  assert.equal(result.resolution.profiles.toArray().length,2);
  const partial=await resolveBuild(plan,[first]);
  assert.deepEqual(partial.resolution.missing.toArray(),[secondId]);
  const empty=await resolveBuild(plan,[]);
  assert.deepEqual(empty.resolution.missing.toArray(),[firstId,secondId].sort());
});

test('tampered, duplicate and extra profiles cannot satisfy an unchanged pin',async()=>{
  assert.deepEqual(await resolveBuild(plan,[first,first]),{ok:false,error:'duplicate_identifier'});
  assert.deepEqual(await resolveBuild(plan,[first+' ']),{ok:false,error:'noncanonical'});
  const changed=first.replace('"max":1010','"max":1011');
  assert.deepEqual(await resolveBuild(plan,[changed,second]),{ok:false,error:'unreferenced_profile'});
  assert.deepEqual(await resolveBuild(plan,[{identity:firstId,canonical:second}]),{ok:false,error:'invalid_document'});
});

test('shared budgets refuse types, sparse arrays, count, per-document and total bytes',async()=>{
  for(const profiles of [null,{},[undefined],new Array(1)]) assert.deepEqual(await resolveBuild(plan,profiles),{ok:false,error:'invalid_document'});
  assert.deepEqual(await resolveBuild(null,[]),{ok:false,error:'invalid_document'});
  assert.deepEqual(await resolveBuild(plan,Array(65).fill('')),{ok:false,error:'invalid_count'});
  assert.deepEqual(await resolveBuild(plan,[' '.repeat(262145)]),{ok:false,error:'too_large'});
  assert.deepEqual(await resolveBuild(plan,Array(16).fill(' '.repeat(262144))),{ok:false,error:'too_large'});
  assert.deepEqual(await resolveBuild(plan,['é'.repeat(131073)]),{ok:false,error:'too_large'});
});

test('crypto loss fails closed even when no profile body was supplied',async()=>{
  const original=Object.getOwnPropertyDescriptor(globalThis,'crypto');
  try{
    Object.defineProperty(globalThis,'crypto',{value:undefined,configurable:true});
    assert.deepEqual(await resolveBuild(plan,[]),{ok:false,error:'crypto_unavailable'});
  }finally{Object.defineProperty(globalThis,'crypto',original);}
});
